blockchain/stake/staketx.go

```diff
@@ -854,8 +987,40 @@ func CheckSSGen(tx *wire.MsgTx) error {
 	// valid push length.
 	if !(firstOutputScriptPrefix[0] == validSSGenVoteOutMinPrefix[0]) ||
 		!pushLengthValid {
-		return stakeRuleError(ErrSSGenBadVotePush, "Second SSGen output "+
-			"had an invalid prefix")
+		return nil, stakeRuleError(ErrSSGenBadVotePush,
+			"Second SSGen output had an invalid prefix")
+	}
+
+	// Check to see if the last output is an OP_RETURN followed by a 2 byte
+	// data push that designates what the data is that follows the
+	// discriminator. In the case of 'T','V' the next data push should be N
+	// hashes. If it is we need to decrease the count on OP_SSGEN tests by
+	// one. This check is only valid if the treasury agenda is active,
+	txOutLen := len(tx.TxOut)
+	lastTxOut := tx.TxOut[len(tx.TxOut)-1]
+	var votes []TreasuryVoteTuple
+	if isTreasuryEnabled &&
+		txscript.GetScriptClass(lastTxOut.Version, lastTxOut.PkScript,
+			isTreasuryEnabled) == txscript.NullDataTy {
+		txOutLen--
+
+		// We call this function in order to prevent rolling of
+		// additional functions that may not conform 100% to consensus.
+		var err error
+		votes, err = GetSSGenTreasuryVotes(lastTxOut.PkScript)
+		if err != nil {
+			return nil, err
+		}
+
+		// If there are votes the TxVersion must be TxVersionTreasury.
+		// This test is done late in order to allow older versions
+		// SSGen if there are no votes.
+		if !(len(votes) > 0 && tx.Version == wire.TxVersionTreasury) {
```

Comment:
Should be `... && tx.Version >= wire.TxVersionTreasury` because later versions will probably be ok, correct? Also the `!` in front makes it really hard for me to read this. Could just do != if we want the logic the same I think.

To sum it up I think `if len(votes) > 0 && tx.Version < wire.TxVersionTreasury {` is better...

Also, in the comment, the version could be `TxVersionTreasury` or higher, correct?

Reply:
After long debates with @davecgh we decided to go with the exact version. This makes upgrading in the future easier since we simply will have a new function for the new version instead of shoehorning it in prior versions.

The ! in front of a positive case is a master of taste. I prefer `if !(total positive case)` over if negative and/or negative. You guys are welcome to change this but I am not going to modify this PR for that.